File: src/model/typed/TypedRetryPolicy.ts
/**
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

import { TypedData } from './TypedData';
import {
	ConstantRetryPolicy,
	ExponentialRetryPolicy,
	NoRetryPolicy,
	RetryPolicy,
} from '../../operations/RetryPolicy';

class RetryPolicyData {
	maxRetries?: number;
	delayInSeconds?: number;
	initialDelayInSeconds?: number;
	maximumDelayInSeconds?: number;
	multiplier?: number;

	constructor(
		maxRetries?: number,
		delayInSeconds?: number,
		initialDelayInSeconds?: number,
		maximumDelayInSeconds?: number,
		multiplier?: number
	) {
		this.maxRetries = maxRetries;
		this.delayInSeconds = delayInSeconds;
		this.initialDelayInSeconds = initialDelayInSeconds;
		this.maximumDelayInSeconds = maximumDelayInSeconds;
		this.multiplier = multiplier;
	}
}

export abstract class TypedRetryPolicy extends TypedData<RetryPolicyData> {
	abstract data: RetryPolicyData;

	static create(retryPolicy: RetryPolicy): TypedRetryPolicy {
		if (retryPolicy instanceof NoRetryPolicy) {
			return new TypedNoRetryPolicy();
		} else if (retryPolicy instanceof ConstantRetryPolicy) {
			return new TypedConstantRetryPolicy(retryPolicy as ConstantRetryPolicy);
		} else if (retryPolicy instanceof ExponentialRetryPolicy) {
			return new TypedExponentialRetryPolicy(retryPolicy as ExponentialRetryPolicy);
		} else {
			throw new Error(`Unknown retry policy (${retryPolicy.constructor.name}).`);
		}
	}
}

export class TypedNoRetryPolicy extends TypedRetryPolicy {
	type = 'NoRetryPolicy';
	data: RetryPolicyData;

	constructor() {
		super();
		this.data = new RetryPolicyData();
	}
}

export class TypedConstantRetryPolicy extends TypedRetryPolicy {
	type = 'ConstantRetryPolicy';
	data: RetryPolicyData;

	constructor(wrapped: ConstantRetryPolicy) {
		super();
		this.data = new RetryPolicyData(wrapped.maxRetries, wrapped.delayInSeconds);
	}
}

export class TypedExponentialRetryPolicy extends TypedRetryPolicy {
	type = 'ExponentialRetryPolicy';
	data: RetryPolicyData;

	constructor(wrapped: ExponentialRetryPolicy) {
		super();
		this.data = new RetryPolicyData(
			wrapped.maxRetries,
			undefined,
			wrapped.initialDelayInSeconds,
			wrapped.maximumDelayInSeconds,
			wrapped.multiplier
		);
	}
}

File: ios/Model/Typed/TypedRetryPolicy.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct TypedRetryPolicy: Decodable {
	// MARK: Properties

	let wrapped: RetryPolicy

	// MARK: Decodable

	init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: TypedCodingKeys.self)
		let type = try container.decode(String.self, forKey: .type)
		switch type {
		case RetryPolicy.CodingKeys.Case.noRetry.stringValue:
			wrapped = .noRetry
		case RetryPolicy.CodingKeys.Case.constant.stringValue:
			let dataContainer = try container.nestedContainer(keyedBy: RetryPolicy.CodingKeys.Constant.self,
															  forKey: .data)
			let maxRetries = try dataContainer.decode(Int.self, forKey: .maxRetries)
			let delayInSeconds = try dataContainer.decode(TimeInterval.self, forKey: .delayInSeconds)
			wrapped = .constant(maxRetries: maxRetries,
								delayInSeconds: delayInSeconds)
		case RetryPolicy.CodingKeys.Case.exponential.stringValue:
			let dataContainer = try container.nestedContainer(keyedBy: RetryPolicy.CodingKeys.Exponential.self,
															  forKey: .data)
			let maxRetries = try dataContainer.decode(Int.self, forKey: .maxRetries)
			let initialDelayInSeconds = try dataContainer.decode(TimeInterval.self, forKey: .initialDelayInSeconds)
			let multiplier = try dataContainer.decode(Double.self, forKey: .multiplier)
			let maximumDelayInSeconds = try dataContainer.decode(TimeInterval.self, forKey: .maximumDelayInSeconds)
			wrapped = .exponential(maxRetries: maxRetries,
								   initialDelayInSeconds: initialDelayInSeconds,
								   multiplier: multiplier,
								   maximumDelayInSeconds: maximumDelayInSeconds)
		default:
			throw DecodingError.dataCorrupted(
				DecodingError.Context(
					codingPath: container.codingPath,
					debugDescription: "Failed to decode retry policy!"
				)
			)
		}
	}
}

extension RetryPolicy {
	enum CodingKeys {
		enum Case: CodingKey {
			case noRetry
			case constant
			case exponential
		}

		enum Constant: CodingKey {
			case maxRetries
			case delayInSeconds
		}

		enum Exponential: CodingKey {
			case maxRetries
			case initialDelayInSeconds
			case multiplier
			case maximumDelayInSeconds
		}
	}
	
	var caseName: String {
		return Mirror(reflecting: self).children.first?.label ?? String(describing: self)
	}
}
